# R.Queen Applied Data Science Capstone - Week 3 Project

<hr />

*Import all needed libraries*

In [194]:
import pandas as pd
import numpy as np
import json
import requests
from IPython.display import display
#!conda install -c conda-forge folium=0.5.0 --yes
import folium
#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors
print('Packages imported...')

Packages imported...


## Part I - Scrape Wikepedia page to get Toronto neighborhoods and create pandas dataframe from results

### *Pull in all raw data from wikepedia*

In [41]:
url = 'http://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(url)[0]
columns = ['PostalCode','Borough','Neighborhood']
df.columns = columns
print(f'The dataset contains {df.shape[0]} rows and {df.shape[1]} columns.')
df.head()

The dataset contains 287 rows and 3 columns.


PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

### *Clean up data*
1. Drop 'not assigned' Boroughs
2. Name 'not assigned' Neighborhoods to Borough name
3. Group by PostCode and combine Neighborhood names

In [42]:
# Drop all 'not assigned' Boroughs
df = df[df['Borough'] != 'Not assigned']
print(df.shape)
df.head(10)

(210, 3)


PostalCode           Borough      Neighborhood
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M6A        North York  Lawrence Heights
6         M6A        North York    Lawrence Manor
7         M7A      Queen's Park      Not assigned
9         M9A      Queen's Park      Queen's Park
10        M1B       Scarborough             Rouge
11        M1B       Scarborough           Malvern
13        M3B        North York   Don Mills North

In [43]:
# Name 'not assigned' Neighborhoods to Borough name
for index, row in df.iterrows():
    if row['Neighborhood'] == 'Not assigned':
        row['Neighborhood'] = row['Borough']
print(df.shape)
df.head(10)

(210, 3)


PostalCode           Borough      Neighborhood
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M6A        North York  Lawrence Heights
6         M6A        North York    Lawrence Manor
7         M7A      Queen's Park      Queen's Park
9         M9A      Queen's Park      Queen's Park
10        M1B       Scarborough             Rouge
11        M1B       Scarborough           Malvern
13        M3B        North York   Don Mills North

In [44]:
# Group by PostCode and combine Neighborhood names
df = df.groupby(['PostalCode','Borough'])['Neighborhood'].apply(lambda x: ','.join(x)).to_frame().reset_index()
print(df.shape)
df.head(10)

(103, 3)


PostalCode      Borough                                   Neighborhood
0        M1B  Scarborough                                  Rouge,Malvern
1        M1C  Scarborough           Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough                Guildwood,Morningside,West Hill
3        M1G  Scarborough                                         Woburn
4        M1H  Scarborough                                      Cedarbrae
5        M1J  Scarborough                            Scarborough Village
6        M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park
7        M1L  Scarborough                  Clairlea,Golden Mile,Oakridge
8        M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West
9        M1N  Scarborough                     Birch Cliff,Cliffside West

In [45]:
print(f'The cleaned dataframe has {df.shape[0]} rows.')

The cleaned dataframe has 103 rows.


<p><i>...(end of Part I)...</i></p>
<hr />

## Part II - Add Latitude and Longitude to DataFrame

### *Unsuccessful with google API to get lat / lng so using supplied CSV file import*

In [71]:
!wget -q -O 'postalcodes.csv' http://cocl.us/Geospatial_data
print('Data downloaded.')

Data downloaded.


In [75]:
# Read CSV into dataframe
pcdf = pd.read_csv('postalcodes.csv')
pcdf.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
pcdf.head(10)

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476
5        M1J  43.744734 -79.239476
6        M1K  43.727929 -79.262029
7        M1L  43.711112 -79.284577
8        M1M  43.716316 -79.239476
9        M1N  43.692657 -79.264848

### *Merging into existing dataframe to add lat / lng columns*

In [77]:
df = df.merge(pcdf, how='left', on=['PostalCode'])
print(df.shape)
df.head(10)

(103, 5)


PostalCode      Borough                                   Neighborhood  \
0        M1B  Scarborough                                  Rouge,Malvern   
1        M1C  Scarborough           Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough                Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                         Woburn   
4        M1H  Scarborough                                      Cedarbrae   
5        M1J  Scarborough                            Scarborough Village   
6        M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park   
7        M1L  Scarborough                  Clairlea,Golden Mile,Oakridge   
8        M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West   
9        M1N  Scarborough                     Birch Cliff,Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

<p><i>...(end of Part II)...</i></p>
<hr />

## Part III - Explore and Cluster the Neighborhoods

### *Create initial map to ensure proper plotting of neighborhoods*

In [195]:
f = folium.Figure(width=900, height=600)
map = folium.Map(location=[43.70011, -79.4163], zoom_start=11)  #Center over Toronto

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = f'{neighborhood} - {borough}'
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map)  
f.add_child(map)    
display(map)

### *A hidden cell assigns my CLIENT_ID and CLIENT_SECRET*

In [95]:
# The code was removed by Watson Studio for sharing.

### *Create routine to get top 100 venues for each neighborhood*

In [97]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, limit=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [124]:
# Run routine for each neghborhood in Toronto dataframe
tor_venues = getNearbyVenues(names=df['Neighborhood'], latitudes=df['Latitude'], longitudes=df['Longitude'])

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

In [125]:
print(f'The Toronto Venues dataframe has {tor_venues.shape[0]} rows and {tor_venues.shape[1]} columns.')
tor_venues.head(10)

The Toronto Venues dataframe has 2235 rows and 7 columns.


Neighborhood  Neighborhood Latitude  \
0                         Rouge,Malvern              43.806686   
1  Highland Creek,Rouge Hill,Port Union              43.784535   
2       Guildwood,Morningside,West Hill              43.763573   
3       Guildwood,Morningside,West Hill              43.763573   
4       Guildwood,Morningside,West Hill              43.763573   
5       Guildwood,Morningside,West Hill              43.763573   
6       Guildwood,Morningside,West Hill              43.763573   
7       Guildwood,Morningside,West Hill              43.763573   
8       Guildwood,Morningside,West Hill              43.763573   
9       Guildwood,Morningside,West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3              -79.188711                G & G Electronics       43.765309   
4              -79.188711                       Marina Spa       43.766000   
5              -79.188711                 Big Bite Burrito       43.766299   
6              -79.188711            Enterprise Rent-A-Car       43.764076   
7              -79.188711            Woburn Medical Centre       43.766631   
8              -79.188711     Lawrence Ave E & Kingston Rd       43.767704   
9              -79.188711                         Eggsmart       43.767800   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa  
5       -79.190720    Mexican Restaurant  
6       -79.193406   Rental Car Location  
7       -79.192286        Medical Center  
8       -79.189490          Intersection  
9       -79.190466        Breakfast Spot

In [126]:
print('There are {} uniques categories.'.format(len(tor_venues['Venue Category'].unique())))

There are 278 uniques categories.


### *Begin to analyze neighborhoods and reshape data for clustering*

In [135]:
# One hot encoding
tor_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix='', prefix_sep='')

tor_onehot['Neighborhood'] = tor_venues['Neighborhood']

tor_onehot.head()

Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                  0                  0        0                   0   
1                  0                  0        0                   0   
2                  0                  0        0                   0   
3                  0                  0        0                   0   
4                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  \
0                    0             0  ...              0   
1                    0             0  ...              0   
2                    0             0  ...              0   
3                    0             0  ...              0   
4                    0             0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 278 columns]

### *Look at mean frequency of occurrence of each category, by Neighborhood*

In [153]:
tor_grouped = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grouped.head(10)

Neighborhood  Accessories Store  \
0                             Adelaide,King,Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                0.0   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...                0.0   
4                              Alderwood,Long Branch                0.0   
5      Bathurst Manor,Downsview North,Wilson Heights                0.0   
6                                    Bayview Village                0.0   
7                   Bedford Park,Lawrence Manor East                0.0   
8                                        Berczy Park                0.0   
9                         Birch Cliff,Cliffside West                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   
5                0.0      0.0                 0.0           0.0   
6                0.0      0.0                 0.0           0.0   
7                0.0      0.0                 0.0           0.0   
8                0.0      0.0                 0.0           0.0   
9                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.02   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   
3             0.0              0.0               0.0                 0.00   
4             0.0              0.0               0.0                 0.00   
5             0.0              0.0               0.0                 0.00   
6             0.0              0.0               0.0                 0.00   
7             0.0              0.0               0.0                 0.04   
8             0.0              0.0               0.0                 0.00   
9             0.0              0.0               0.0                 0.00   

   ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...            0.0                       0.020000               0.0   
1  ...            0.0                       0.000000               0.0   
2  ...            0.0                       0.000000               0.0   
3  ...            0.0                       0.000000               0.0   
4  ...            0.0                       0.000000               0.0   
5  ...            0.0                       0.000000               0.0   
6  ...            0.0                       0.000000               0.0   
7  ...            0.0                       0.000000               0.0   
8  ...            0.0                       0.017857               0.0   
9  ...            0.0                       0.000000               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0      0.01          0.0   
1          0.0                    0.0              0.0      0.00          0.0   
2          0.0                    0.0              0.0      0.00          0.0   
3          0.0                    0.0              0.0      0.00          0.0   
4          0.0                    0.0              0.0      0.00          0.0   
5          0.0                    0.0              0.0      0.00          0.0   
6          0.0                    0.0              0.0      0.00          0.0   
7          0.0                    0.0              0.0      0.00          0.0   
8          0.0                    0.0              0.0      0.00          0.0   
9  

### *Add top 10 Venues to a dataframe for each Neighborhood*

In [138]:
def return_top_venues(row, num):
    row_cat = row.iloc[1:]
    row_cat_sort = row_cat.sort_values(ascending=False)
    return row_cat_sort.index.values[0:num]

In [154]:
num = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

venues_sort = pd.DataFrame(columns=columns)
venues_sort['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    venues_sort.iloc[ind, 1:] = return_top_venues(tor_grouped.iloc[ind, 1:], num)
    
venues_sort.head(10)

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                  Park   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...           Pizza Place   
4                              Alderwood,Long Branch           Pizza Place   
5      Bathurst Manor,Downsview North,Wilson Heights           Coffee Shop   
6                                    Bayview Village                  Café   
7                   Bedford Park,Lawrence Manor East    Italian Restaurant   
8                                        Berczy Park           Coffee Shop   
9                         Birch Cliff,Cliffside West       College Stadium   

   2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0                   Café                 Steakhouse       Thai Restaurant   
1         Breakfast Spot  Latin American Restaurant          Skating Rink   
2             Playground                Yoga Studio             Drugstore   
3               Pharmacy                 Beer Store   Fried Chicken Joint   
4            Coffee Shop               Skating Rink              Pharmacy   
5          Shopping Mall                Pizza Place            Restaurant   
6                   Bank        Japanese Restaurant    Chinese Restaurant   
7            Coffee Shop                   Boutique            Hobby Shop   
8     Seafood Restaurant             Farmers Market                  Café   
9  General Entertainment               Skating Rink                  Café   

  5th Most Common Venue 6th Most Common Venue    7th Most Common Venue  \
0                Bakery      Asian Restaurant             Burger Joint   
1   Dumpling Restaurant        Discount Store                  Dog Run   
2    Dim Sum Restaurant                 Diner           Discount Store   
3  Fast Food Restaurant        Sandwich Place            Grocery Store   
4                  Pool                   Pub           Sandwich Place   
5           Supermarket      Sushi Restaurant                     Bank   
6    Dim Sum Restaurant        Discount Store                  Dog Run   
7        Sandwich Place        Breakfast Spot  Comfort Food Restaurant   
8           Cheese Shop          Cocktail Bar                 Beer Bar   
9          Dessert Shop    Dim Sum Restaurant                    Diner   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Bar            Restaurant       Sushi Restaurant  
1      Doner Restaurant            Donut Shop              Drugstore  
2               Dog Run      Doner Restaurant             Donut Shop  
3      Airport Terminal          Dessert Shop     Falafel Restaurant  
4                   Gym   American Restaurant       Department Store  
5           Bridal Shop  Fast Food Restaurant                  Diner  
6      Doner Restaurant            Donut Shop              Drugstore  
7  Fast Food Restaurant             Juice Bar                Butcher  
8                Bakery            Steakhouse    Japanese Restaurant  
9        Discount Store               Dog Run       Doner Restaurant

### *Cluster Neighborhoods based on similar venue attributes*

In [155]:
k = 5
tor_clusters = tor_grouped.drop('Neighborhood', 1)

# K-Means Clustering
kmeans = KMeans(n_clusters=k, random_state=4)
kmeans.fit(tor_clusters)

kmeans.labels_[0:10]

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [156]:
# Add cluster label to dataframe of neighborhoods

venues_sort.insert(0, 'ClusterLabel', kmeans.labels_)

tor_merged = df
tor_merged = tor_merged.join(venues_sort.set_index('Neighborhood'), on='Neighborhood')

tor_merged.head(10)

PostalCode      Borough                                   Neighborhood  \
0        M1B  Scarborough                                  Rouge,Malvern   
1        M1C  Scarborough           Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough                Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                         Woburn   
4        M1H  Scarborough                                      Cedarbrae   
5        M1J  Scarborough                            Scarborough Village   
6        M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park   
7        M1L  Scarborough                  Clairlea,Golden Mile,Oakridge   
8        M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West   
9        M1N  Scarborough                     Birch Cliff,Cliffside West   

    Latitude  Longitude  ClusterLabel 1st Most Common Venue  \
0  43.806686 -79.194353           1.0  Fast Food Restaurant   
1  43.784535 -79.160497           0.0                   Bar   
2  43.763573 -79.188711           0.0        Medical Center   
3  43.770992 -79.216917           0.0           Coffee Shop   
4  43.773136 -79.239476           0.0      Hakka Restaurant   
5  43.744734 -79.239476           0.0            Playground   
6  43.727929 -79.262029           0.0      Department Store   
7  43.711112 -79.284577           0.0              Bus Line   
8  43.716316 -79.239476           0.0                 Motel   
9  43.692657 -79.264848           0.0       College Stadium   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Yoga Studio             Drugstore    Dim Sum Restaurant   
1            Yoga Studio   Dumpling Restaurant        Discount Store   
2    Rental Car Location                   Spa        Breakfast Spot   
3      Korean Restaurant           Yoga Studio   Dumpling Restaurant   
4            Gas Station                  Bank   Fried Chicken Joint   
5            Yoga Studio             Drugstore    Dim Sum Restaurant   
6         Discount Store           Coffee Shop            Hobby Shop   
7                 Bakery          Soccer Field                  Park   
8    American Restaurant             Drugstore    Dim Sum Restaurant   
9  General Entertainment          Skating Rink                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Diner        Discount Store               Dog Run   
1               Dog Run      Doner Restaurant            Donut Shop   
2           Pizza Place     Electronics Store          Intersection   
3                 Diner        Discount Store               Dog Run   
4  Caribbean Restaurant    Athletics & Sports       Thai Restaurant   
5                 Diner        Discount Store               Dog Run   
6     Convenience Store   Dumpling Restaurant                 Diner   
7  Fast Food Restaurant          Intersection           Bus Station   
8                 Diner        Discount Store               Dog Run   
9          Dessert Shop    Dim Sum Restaurant                 Diner   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                   Donut Shop    Dumpling Restaurant  
1             Drugstore  Eastern European Restaurant     Dim Sum Restaurant  
2    Mexican Restaurant             Doner Restaurant         Discount Store  
3      Doner Restaurant                   Donut Shop              Drugstore  
4                Bakery                    Drugstore             Donut Shop  
5      Doner Restaurant                   Donut Shop    Dumpling Restaurant  
6               Dog Run             Doner Restaurant             Donut Shop  
7         Metro Station                      Dog Run       Doner Restaurant  
8      Doner Restaurant                   Donut Shop            Yoga Studio  
9        Discount Store                      Dog Run       Doner Restaurant

In [179]:
tor_merged = tor_merged[np.isfinite(tor_merged['ClusterLabel'])]

### *Visualize new neighborhood clusters*

In [196]:
#f = folium.Figure(width=900, height=600)
map = folium.Map(location=[43.70011, -79.4163], zoom_start=11)  #Center over Toronto

# add markers to map
marker_color = ['red','blue','yellow','green','black']
for lat, lng, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['ClusterLabel']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    col = int(cluster)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color=marker_color[col],
        fill=True,
        fill_color=marker_color[col],
        fill_opacity=0.7).add_to(map)  
#f.add_child(map)    
display(map)

### *Inspect Clusters to see defining categories*

In [193]:
tor_merged['ClusterLabel'].value_counts(normalize=True)

0.0    0.851485
2.0    0.099010
3.0    0.029703
4.0    0.009901
1.0    0.009901
Name: ClusterLabel, dtype: float64

### Cluster 0

In [181]:
tor_merged[tor_merged['ClusterLabel'] == 0]

PostalCode           Borough  \
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
15         M1W       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
22         M2N        North York   
24         M2R        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
31         M3L        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
..         ...               ...   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
92         M8Z         Etobicoke   
93         M9A      Queen's Park   
95         M9C         Etobicoke   
96         M9L        North York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497   
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711   
3                                               Woburn  43.770992 -79.216917   
4                                            Cedarbrae  43.773136 -79.239476   
5                                  Scarborough Village  43.744734 -79.239476   
6            East Birchmount Park,Ionview,Kennedy Park  43.727929 -79.262029   
7                        Clairlea,Golden Mile,Oakridge  43.711112 -79.284577   
8        Cliffcrest,Cliffside,Scarborough Village West  43.716316 -79.239476   
9                           Birch Cliff,Cliffside West  43.692657 -79.264848   
10   Dorset Park,Scarborough Town Centre,Wexford He...  43.757410 -79.273304   
11                                    Maryvale,Wexford  43.750072 -79.295849   
12                                           Agincourt  43.794200 -79.262029   
13               Clarks Corners,Sullivan,Tam O'Shanter  43.781638 -79.304302   
15                                     L'Amoreaux West  43.799525 -79.318389   
17                                   Hillcrest Village  43.803762 -79.363452   
18                          Fairview,Henry Farm,Oriole  43.778517 -79.346556   
19                                     Bayview Village  43.786947 -79.385975   
20                             Silver Hills,York Mills  43.757490 -79.374714   
22                                    Willowdale South  43.770120 -79.408493   
24                                     Willowdale West  43.782736 -79.442259   
26                                     Don Mills North  43.745906 -79.352188   
27         

### Cluster 1

In [182]:
tor_merged[tor_merged['ClusterLabel'] == 1]

PostalCode      Borough   Neighborhood   Latitude  Longitude  ClusterLabel  \
0        M1B  Scarborough  Rouge,Malvern  43.806686 -79.194353           1.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Fast Food Restaurant           Yoga Studio             Drugstore   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0               Dog Run      Doner Restaurant            Donut Shop   

  10th Most Common Venue  
0    Dumpling Restaurant

### Cluster 2

In [183]:
tor_merged[tor_merged['ClusterLabel'] == 2]

PostalCode           Borough  \
14        M1V       Scarborough   
23        M2P        North York   
25        M3A        North York   
30        M3K        North York   
40        M4J         East York   
50        M4W  Downtown Toronto   
74        M6E              York   
79        M6L        North York   
90        M8X         Etobicoke   
98        M9N              York   

                                         Neighborhood   Latitude  Longitude  \
14  Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577   
23                                    York Mills West  43.752758 -79.400049   
25                                          Parkwoods  43.753259 -79.329656   
30                         CFB Toronto,Downsview East  43.737473 -79.464763   
40                                       East Toronto  43.685347 -79.338106   
50                                           Rosedale  43.679563 -79.377529   
74                                Caledonia-Fairbanks  43.689026 -79.453512   
79                   Downsview,North Park,Upwood Park  43.713756 -79.490074   
90        The Kingsway,Montgomery Road,Old Mill North  43.653654 -79.506944   
98                                             Weston  43.706876 -79.518188   

    ClusterLabel 1st Most Common Venue 2nd Most Common Venue  \
14           2.0                  Park            Playground   
23           2.0                  Park                  Bank   
25           2.0                  Park     Food & Drink Shop   
30           2.0               Airport                  Park   
40           2.0                  Park           Coffee Shop   
50           2.0                  Park            Playground   
74           2.0                  Park         Women's Store   
79           2.0                Bakery                  Park   
90           2.0                 River                  Park   
98           2.0                  Park     Convenience Store   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14                 Yoga Studio             Drugstore    Dim Sum Restaurant   
23           Convenience Store           Yoga Studio   Dumpling Restaurant   
25                 Yoga Studio            Donut Shop    Dim Sum Restaurant   
30                 Yoga Studio   Dumpling Restaurant                 Diner   
40           Convenience Store           Yoga Studio             Drugstore   
50                       Trail           Yoga Studio            Donut Shop   
74                      Market  Fast Food Restaurant          Gourmet Shop   
79  Construction & Landscaping           Yoga Studio   Dumpling Restaurant   
90                        Pool            Donut Shop          Dessert Shop   
98                 Yoga Studio             Drugstore    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14                 Diner        Discount Store               Dog Run   
23        Discount Store               Dog Run      Doner Restaurant   
25                 Diner        Discount Store               Dog Run   
30        Discount Store               Dog Run      Doner Restaurant   
40                 Diner        Discount Store               Dog Run   
50          Dessert Shop    Dim Sum Restaurant                 Diner   
74           Event Space  Ethiopian Restaurant   Empanada Restaurant   
79        Discount Store               Dog Run      Doner Restaurant   
90    Dim Sum Restaurant                 Diner        Discount Store   
98                 Diner        Discount Store               Dog Run   

   9th Most Common Venue       10th Most Common Venue  
14      Doner Restaurant                   Donut Shop  
23            Donut Shop                    Drugstore  
25      Doner Restaurant                    Drugstore  
30            Donut Shop                    Drugstore  
40      Doner Restaurant                   Donut Shop  
50        Discount Store                      Dog Run 

### Cluster 3

In [184]:
tor_merged[tor_merged['ClusterLabel'] == 3]

PostalCode     Borough                                       Neighborhood  \
32        M3M  North York                                  Downsview Central   
91        M8Y   Etobicoke  Humber Bay,King's Mill Park,Kingsway Park Sout...   
97        M9M  North York                                    Emery,Humberlea   

     Latitude  Longitude  ClusterLabel 1st Most Common Venue  \
32  43.728496 -79.495697           3.0        Baseball Field   
91  43.636258 -79.498509           3.0         Deli / Bodega   
97  43.724766 -79.532242           3.0        Baseball Field   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32            Food Truck           Yoga Studio             Drugstore   
91        Baseball Field           Yoga Studio    Dim Sum Restaurant   
97           Yoga Studio   Dumpling Restaurant        Discount Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32        Discount Store               Dog Run      Doner Restaurant   
91        Discount Store               Dog Run      Doner Restaurant   
97               Dog Run      Doner Restaurant            Donut Shop   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
32            Donut Shop          Dumpling Restaurant     Dim Sum Restaurant  
91            Donut Shop                    Drugstore    Dumpling Restaurant  
97             Drugstore  Eastern European Restaurant     Dim Sum Restaurant

### Cluster 4

In [185]:
tor_merged[tor_merged['ClusterLabel'] == 4]

PostalCode    Borough                                       Neighborhood  \
94        M9B  Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...   

     Latitude  Longitude  ClusterLabel 1st Most Common Venue  \
94  43.650943 -79.554724           4.0   Filipino Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
94           Yoga Studio   Dumpling Restaurant                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
94        Discount Store               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
94            Donut Shop             Drugstore  Eastern European Restaurant

## Assessment

<p>
    Although I was able to successfully pull in all data, transform, and cluster the Toronto neighborhoods based on venue types, this feature of the neighborhoods does not appear to be the strongest indicator for clustering, given that 85% of the data fell into a single cluster.
    <br /><br />
    At 3% of the data, Cluster 3 appears to be the baseball fields in the area.
    <br /><br />
    At 10% of the data, Cluster 2 appears to be many of the parks and playgrounds.
    <br /><br />
    At 1% each, Clusters 1 and 4 do not give enough information to arrive at defining characteristics.
    <br /><br /><br />
    In running this project again, I would pick a different feature of these neighborhoods with which to cluster, such as population demographic statistics.
</p>

<p><i>...(end of Part III)...</i></p>
<hr />

## End of Assignment - Thank You